$Id: djbdns-dnscache-sigpipe.patch 295 2007-09-26 01:12:23Z geo $

List:       djbdns
Subject:    Re: dnscache exiting on SIGNAL 13
From:       Mark Delany <n5xbg-aanea () qmda ! emu ! st>
Date:       2003-03-19 2:08:17

dnscache is answering a tcp query from 203.134.188.70 and
on writing the reply back to the client, dnscache gets a SIGPIPE
because the querying client has closed the socket.

I don't see any code that protects dnscache from SIGPIPEs so
as a stopgap I'd be tempted to ignore the signal. Here's an untested
patch:

*** dnscache.c.orig     Sun Feb 11 13:11:45 2001
--- dnscache.c  Tue Mar 18 17:22:03 2003
***************
*** 1,4 ****
--- 1,5 ----
  #include <unistd.h>
+ #include <signal.h>
  #include "env.h"
  #include "exit.h"
  #include "scan.h"
***************
*** 391,396 ****
--- 392,398 ----
    char *x;
    unsigned long cachesize;
  
+   signal(SIGPIPE, SIG_IGN);
    x = env_get("IP");
    if (!x)
      strerr_die2x(111,FATAL,"$IP not set");
